src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/DeserializationHelpers.cs

```diff
@@ -235,6 +262,74 @@ internal static void SetProperty(object resource, string propertyName, object va
             }
         }
 
+        internal static void SetInstanceAnnotations(ODataResource oDataResource, IODataInstanceAnnotationContainer instanceAnnotationContainer, 
+            ODataDeserializerProvider deserializerProvider, ODataDeserializerContext readContext)
+        {
+            if(oDataResource.InstanceAnnotations != null && oDataResource.InstanceAnnotations.Count > 0)
+            {
+                foreach (ODataInstanceAnnotation annotation in oDataResource.InstanceAnnotations)
+                {
+                    AddInstanceAnnotationToContainer(instanceAnnotationContainer, deserializerProvider, readContext, annotation,string.Empty);
+                }
+            }
+
+            foreach(ODataProperty property in oDataResource.Properties)
+            {
+                if(property.InstanceAnnotations != null && property.InstanceAnnotations.Count > 0)
+                {
+                    foreach (ODataInstanceAnnotation annotation in property.InstanceAnnotations)
+                    {
+                        AddInstanceAnnotationToContainer(instanceAnnotationContainer, deserializerProvider, readContext, annotation, property.Name);
+                    }
+                }
+            }
+        }
+
+        private static void AddInstanceAnnotationToContainer(IODataInstanceAnnotationContainer instanceAnnotationContainer, ODataDeserializerProvider deserializerProvider, 
+            ODataDeserializerContext readContext, ODataInstanceAnnotation annotation, string propertyName)
+        {
+            EdmTypeKind type;
+            IEdmTypeReference propertyType = null;
+
+            object annotationValue = ConvertValue(annotation.Value, ref propertyType, deserializerProvider, readContext, out type);
+
+            if (string.IsNullOrEmpty(propertyName))
+            {
+                instanceAnnotationContainer.AddAnnotation(annotation.Name, annotationValue);
+            }
+            else
+            {
+                instanceAnnotationContainer.AddPropertyAnnotation(propertyName,annotation.Name, annotationValue);
+            }
+        }
+
+        public static IODataInstanceAnnotationContainer GetAnnotationContainer(PropertyInfo propertyInfo, object value, object resource, IDelta delta)
+        {
+            IODataInstanceAnnotationContainer instanceAnnotationContainer = value as IODataInstanceAnnotationContainer;
+
+            if (instanceAnnotationContainer == null)
+            {
+                try
+                {
+                    instanceAnnotationContainer = Activator.CreateInstance(propertyInfo.PropertyType) as IODataInstanceAnnotationContainer;
```

Comment:
Consider special casing the common scenario of (propertyInfo.PropertyType == ODataInstanceAnnotationContainer || propertyInfo.PropertyType == IODataInstanceAnnotationContainer) and directly call new ODataInstanceAnnotationContainer(), rather than going through reflection to create the instance for that scenario.